### ISSUE TITLE
CORS does not work with web sockets

### ISSUE THREAD

**Issue by moorecode** [unosquare/embedio#358]
**Describe the bug**
The CORS headers are not added to websocket requests.

**To Reproduce**
``` c#
server = new WebServer(o => o
                    .WithUrlPrefix("http://0.0.0.0:5000/")
                    .WithMode(HttpListenerMode.EmbedIO)
                )
                .WithCors()
                .WithModule(<WebSocketModule>);
```
1. Start web server
2. try to connect using LAN ip
3. connections fails (CORS error)

**Expected behavior**
the `.WithCors()` should allow resource sharing on websocket connections

**Desktop (please complete the following information):**
 - OS: Windows 12
 - Browser: Chrome, firefox
 - Version : embedio 3.0.5

**Comment by rdeago** [comment]
Hello @moorecode, thanks for your input.

I have identified the problem and am working on a fix.

The root cause, in case you're interested, is that WebSocketModule does not use the HTTP context's response object (`EmbedIO.NetInternal.HttpListenerResponse`), whose headers are in fact correctly set by `CorsModule`: instead, a new "quasi-response" object (`EmbedIO.NetInternal.HttpResponse`, probably a remnant of WebSocketSharp code) is created and sent to the client. Headers from the first object are not passed on to the second one, nor is the original response object aware that headers have already been sent by the time a `WebSocketContext` is created.

**Comment by rdeago** [comment]
@moorecode I think I have fixed the issue. Sort of.

But now I wonder, was it an issue to begin with?

[RFC6455](https://tools.ietf.org/html/rfc6455#section-1.3) does not mention any response headers besides the ones strictly needed for the WebSocket opening handshake. This does not necessarily mean that other headers are prohibited, or even discouraged, but it made me wondering; so I googled a bit and found [this article](https://blog.securityevaluators.com/websockets-not-bound-by-cors-does-this-mean-2e7819374acc), according to which WebSocket requests are not subject to Cross-Origin Policy and therefore do not need CORS headers.

Even if this issue is apparently fixed, it would seem that the most correct course of action is to add WebSocket modules to a `WebServer` _before_ the CORS module, despite the documentation stating otherwise.

@geoperez this matter probably deserves more research. Shame on me for not doing it _before_ diving into code! I also did not add a regression test: not because I forgot, but because I didn't find a way to get the HTTP response headers from `ClientWebSocket`.

**Comment by moorecode** [comment]
Hi @rdeago, thanks so much for your work! This was strange to me as well as I though the same as you have discovered. Though I seem to have compelling evidence that the browsers I have tested (Chrome and Firefox) do appear to block the connections from different origins. When initialising a web server with the url prefix `http://0.0.0.0:5000` I could not connect at all, though using the devices IP address or `localhost` and ensuring the url I used matched, I was able to connect to the web socket.

**Comment by rdeago** [comment]
Initializing the web server with a URL is a bad habit we had to inherit from Microsoft's `HttpListener`. Its purpose is (or should be) to be able to bind a HTTP listener even to a specific path, like `http://localhost:5000/just_here/`, and/or to one or more DNS names (assuming Windows Server + Active Directory).

EmbedIO's internal, Mono-derived `HttpListener`, not being a managed interface to the `Http.sys` Windows driver, just tries to emulate its behavior for at least the simplest use cases. I sincerely don't know that part of code well enough, but I suspect that 0.0.0.0 as an IP address for a listener may be kind of an edge case.

What happens if you use `http://+:5000` as an URL for your web server?

Also, I'm getting more and more uncomfortable with the idea of including CORS headers (or any header not specified in [RFC6455, section 4.2.2](https://tools.ietf.org/html/rfc6455#section-4.2.2)) in the WebSocket opening handshake. Although, from what I understand so far, the RFC does not explicitly prohibit them, some clients could follow a strict interpretation of the RFC and fail the connection _because_ of the presence of non-RFC-listed headers in an otherwise valid `101 Switching Protocols` response.

I can also confirm that CORS was not the problem to start with. Browsers do not care about same-origin policy when it comes to WebSockets, so CORS headers (normally used to mitigate that policy) are, at best, ignored.

@moorecode can you post more details about the "CORS error" you mentioned in the original post? I'm pretty sure the matter has to be investigated again from a different angle.

**Comment by moorecode** [comment]
I suppose I may have been hasty in categorising the error as a CORS error in the first place, the actual error in both firefox and chrome indicate some kind of malformed response from the server. When the addresses do not match (i.e running the serving with `0.0.0.0` or `localhost` and connecting from `192.168...`) the Chrome error reads:  

> WebSocket connection to 'ws://192.168...:5000/socket' failed: Error during WebSocket handshake: Unexpected response code: 200

And Firefox:
> Firefox can’t establish a connection to the server at ws://192.168...:5000/socket.


Though when I used `http://+:5000` to initialise the server, everything works fine. What kind of behaviour do we expect from the use of `+` in this context? The same as `0.0.0.0`?


In summary I think you are right in saying that CORS has nothing to do this issue, though it does seem strange that binding to a specific address would cause connection errors/malformed responses.
